File: modules/vescape-core/ios/diagnostics/ConnectionTrace.swift
import Foundation

/// Connection trace contract (ADR 0035). One correlated workflow spans lifecycle, scanner,
/// connection, service, and recording code, so every layer emits child events under the same
/// `workflow_id` instead of minting its own correlation.
///
/// Local Diagnostic Events (ADR 0007) intentionally carry full Board ids and BLE ids; they stay on
/// device and in platform logs. Authentication data, PINs, tokens, and telemetry payloads are
/// excluded by contract — see `ConnectionTrace.isSensitiveField`.
///
/// @parity /modules/vescape-core/android/src/main/java/expo/modules/vescapecore/diagnostics/ConnectionTrace.kt
/// @parity /src/modules/diagnostics/connectionTrace.ts
internal enum ConnectionTraceEvent {
  static let workflowStarted = "connection_workflow_started"
  static let workflowFinished = "connection_workflow_finished"

  static let presenceScanStarted = "presence_scan_started"
  static let presenceScanReady = "presence_scan_ready"
  static let presenceScanObserved = "presence_scan_observed"
  static let presenceScanMatched = "presence_scan_matched"
  static let presenceScanTimeout = "presence_scan_timeout"
  static let presenceScanCancelled = "presence_scan_cancelled"
  static let presenceScanSkipped = "presence_scan_skipped"
  static let presenceScanFailed = "presence_scan_failed"

  static let ownerGranted = "connection_owner_granted"
  static let ownerDenied = "connection_owner_denied"
  static let ownerReleased = "connection_owner_released"

  static let connectIntentCreated = "connect_intent_created"
  static let connectIntentCleared = "connect_intent_cleared"
  static let autoConnectPromoted = "auto_connect_promoted"
  static let autoConnectSkipped = "auto_connect_skipped"
  static let autoStartArmed = "auto_start_armed"
  static let autoStartTriggered = "auto_start_triggered"
  static let autoStartSkipped = "auto_start_skipped"
  static let alternativeHintOffered = "alternative_hint_offered"
  static let alternativeHintAccepted = "alternative_hint_accepted"
  static let alternativeHintDismissed = "alternative_hint_dismissed"

  static let pauseStarted = "connection_pause_started"
  static let pauseCleared = "connection_pause_cleared"
  static let pauseExpired = "connection_pause_expired"
  static let pauseBlocked = "connection_pause_blocked"

  static let serviceStarted = "connection_service_started"
  static let servicePromotedForeground = "connection_service_promoted_foreground"
  static let serviceDemotedBackground = "connection_service_demoted_background"
  static let serviceStopped = "connection_service_stopped"

  static let boardSelected = "board_selected"
  static let boardLinkPersisted = "board_link_persisted"
  static let boardLinkFailed = "board_link_failed"

  static let rideSummaryPrepared = "ride_summary_prepared"
  static let rideSummaryNotified = "ride_summary_notified"
  static let rideSummarySkipped = "ride_summary_skipped"
}

/// Who owns connection work, in the precedence order of ADR 0035.
internal enum ConnectionTraceOwner {
  static let boardSession = "board_session"
  static let connectIntent = "connect_intent"
  static let autoStart = "auto_start"
  static let autoConnect = "auto_connect"
  static let alternativeHint = "alternative_hint"
  static let addBoardScan = "add_board_scan"
  static let boardProbe = "board_probe"
  static let none = "none"
}

/// Why the workflow exists. Set once at `ConnectionTrace.start`.
internal enum ConnectionTraceOrigin {
  static let foregroundEntry = "foreground_entry"
  static let explicitConnect = "explicit_connect"
  static let autoStartWake = "auto_start_wake"
  static let alternativeHintSwitch = "alternative_hint_switch"
  static let addBoardScan = "add_board_scan"
  static let boardProbe = "board_probe"
  static let reconnect = "reconnect"
  static let manualDisconnect = "manual_disconnect"
  static let endRide = "end_ride"
  static let appExit = "app_exit"
  static let taskRemoved = "task_removed"
  static let rideFinalized = "ride_finalized"
}

/// Field names. Later slices reuse these instead of inventing new keys.
internal enum ConnectionTraceField {
  static let workflowId = "workflow_id"
  static let workflowOrigin = "workflow_origin"
  static let workflowOwner = "workflow_owner"
  static let workflowStartedAt = "workflow_started_at"
  static let elapsedMs = "elapsed_ms"

  static let boardId = "board_id"
  static let bleId = "ble_id"
  static let boardNickname = "board_nickname"

  static let decision = "decision"
  static let reason = "reason"
  static let ownerPrevious = "owner_previous"
  static let ownerRequested = "owner_requested"

  static let deadlineMs = "deadline_ms"
  static let deadlineAt = "deadline_at"
  static let attempt = "attempt"

  static let scanPurpose = "scan_purpose"
  static let observationCount = "observation_count"
  static let rssi = "rssi"

  static let pauseSource = "pause_source"
  static let pausedUntil = "paused_until"

  static let autoConnectEnabled = "auto_connect_enabled"
  static let autoStartEnabled = "auto_start_enabled"
  static let bluetoothEnabled = "bluetooth_enabled"
  static let permissionGranted = "permission_granted"
  static let appForeground = "app_foreground"
  static let serviceState = "service_state"

  static let rideId = "ride_id"
  static let platformErrorCode = "platform_error_code"
  static let platformErrorDomain = "platform_error_domain"
}

/// Value of `ConnectionTraceField.decision`.
internal enum ConnectionTraceDecision {
  static let granted = "granted"
  static let denied = "denied"
  static let deferred = "deferred"
  static let skipped = "skipped"
  static let completed = "completed"
  static let timeout = "timeout"
  static let cancelled = "cancelled"
  static let failed = "failed"
}

/// Terminal reason names. Value of `ConnectionTraceField.reason`.
internal enum ConnectionTraceReason {
  static let matched = "matched"
  static let noLinkedBoards = "no_linked_boards"
  static let noBoardLink = "no_board_link"
  static let noSelectedBoard = "no_selected_board"
  static let boardNotPresent = "board_not_present"
  static let bluetoothDisabled = "bluetooth_disabled"
  static let permissionMissing = "permission_missing"
  static let scannerUnavailable = "scanner_unavailable"
  static let scannerBusy = "scanner_busy"
  static let autoConnectDisabled = "auto_connect_disabled"
  static let autoStartDisabled = "auto_start_disabled"
  static let connectionPaused = "connection_paused"
  static let higherPriorityOwner = "higher_priority_owner"
  static let sessionAlreadyActive = "session_already_active"
  static let connectIntentActive = "connect_intent_active"
  static let userCancelled = "user_cancelled"
  static let stopSearch = "stop_search"
  static let deadlineExpired = "deadline_expired"
  static let manualDisconnect = "manual_disconnect"
  static let endRide = "end_ride"
  static let appExit = "app_exit"
  static let taskRemoved = "task_removed"
  static let autoClose = "auto_close"
  static let mechanicalTeardown = "mechanical_teardown"
  static let probeCancelled = "probe_cancelled"
  static let platformError = "platform_error"
}

/// One correlated connection workflow. Pass the handle across layers — lifecycle → scanner →
/// connection → service → recording — so the correlation id survives handoff.
internal final class ConnectionWorkflow {
  let workflowId: String
  let origin: String

  private let reporter: DiagnosticReporter
  private let startedAtMs: Int64
  private let lock = NSLock()
  private var currentOwner: String

  init(reporter: DiagnosticReporter, workflowId: String, origin: String, startedAtMs: Int64, owner: String) {
    self.reporter = reporter
    self.workflowId = workflowId
    self.origin = origin
    self.startedAtMs = startedAtMs
    self.currentOwner = owner
  }

  var owner: String {
    lock.lock()
    defer { lock.unlock() }
    return currentOwner
  }

  /// Record a new owner for subsequent child events. Returns the same handle for chaining.
  @discardableResult
  func handoff(owner: String) -> ConnectionWorkflow {
    lock.lock()
    currentOwner = owner
    lock.unlock()
    return self
  }

  func event(_ eventName: String, fields: [String: Any?] = [:]) {
    reporter.capture(eventName: eventName, properties: baseFields().merging(ConnectionTrace.sanitize(fields)) { _, new in new })
  }

  func finish(decision: String, reason: String, fields: [String: Any?] = [:]) {
    var merged: [String: Any?] = [
      ConnectionTraceField.decision: decision,
      ConnectionTraceField.reason: reason,
    ]
    for (key, value) in fields { merged[key] = value }
    event(ConnectionTraceEvent.workflowFinished, fields: merged)
  }

  private func baseFields() -> [String: Any?] {
    [
      ConnectionTraceField.workflowId: workflowId,
      ConnectionTraceField.workflowOrigin: origin,
      ConnectionTraceField.workflowOwner: owner,
      ConnectionTraceField.workflowStartedAt: startedAtMs,
      ConnectionTraceField.elapsedMs: ConnectionTrace.now() - startedAtMs,
    ]
  }
}

internal enum ConnectionTrace {
  private static let sensitiveMarkers = [
    "auth",
    "credential",
    "jwt",
    "password",
    "payload",
    "pin",
    "secret",
    "session_token",
    "telemetry",
    "token",
  ]

  /// Start a correlated workflow and emit `connection_workflow_started`.
  static func start(
    origin: String,
    owner: String = ConnectionTraceOwner.none,
    fields: [String: Any?] = [:],
    reporter: DiagnosticReporter = .shared
  ) -> ConnectionWorkflow {
    let workflow = ConnectionWorkflow(
      reporter: reporter,
      workflowId: UUID().uuidString,
      origin: origin,
      startedAtMs: now(),
      owner: owner
    )
    workflow.event(ConnectionTraceEvent.workflowStarted, fields: fields)
    return workflow
  }

  /// Rebuild a handle for a workflow that already started, so a layer that only received the
  /// correlation id (background task, notification action, JS intent) keeps emitting under it.
  static func resume(
    workflowId: String,
    origin: String,
    startedAtMs: Int64,
    owner: String = ConnectionTraceOwner.none,
    reporter: DiagnosticReporter = .shared
  ) -> ConnectionWorkflow {
    ConnectionWorkflow(
      reporter: reporter,
      workflowId: workflowId,
      origin: origin,
      startedAtMs: startedAtMs,
      owner: owner
    )
  }

  /// Contract-level exclusion of auth data, PINs, tokens, and telemetry payloads.
  static func isSensitiveField(_ key: String) -> Bool {
    let lowered = key.lowercased()
    return sensitiveMarkers.contains { lowered.contains($0) }
  }

  static func sanitize(_ fields: [String: Any?]) -> [String: Any?] {
    fields.filter { !isSensitiveField($0.key) }
  }

  static func now() -> Int64 {
    Int64(Date().timeIntervalSince1970 * 1000.0)
  }
}

File: modules/vescape-core/android/src/main/java/expo/modules/vescapecore/diagnostics/ConnectionTrace.kt
package expo.modules.vescapecore.diagnostics

import android.content.Context

/**
 * Connection trace contract (ADR 0035). One correlated workflow spans lifecycle, scanner,
 * connection, service, and recording code, so every layer emits child events under the same
 * `workflow_id` instead of minting its own correlation.
 *
 * Local Diagnostic Events (ADR 0007) intentionally carry full Board ids and BLE ids; they stay on
 * device and in platform logs. Authentication data, PINs, tokens, and telemetry payloads are
 * excluded by contract — see [ConnectionTrace.isSensitiveField].
 *
 * @parity /modules/vescape-core/ios/diagnostics/ConnectionTrace.swift
 * @parity /src/modules/diagnostics/connectionTrace.ts
 */
object ConnectionTraceEvent {
    const val WORKFLOW_STARTED = "connection_workflow_started"
    const val WORKFLOW_FINISHED = "connection_workflow_finished"

    const val PRESENCE_SCAN_STARTED = "presence_scan_started"
    const val PRESENCE_SCAN_READY = "presence_scan_ready"
    const val PRESENCE_SCAN_OBSERVED = "presence_scan_observed"
    const val PRESENCE_SCAN_MATCHED = "presence_scan_matched"
    const val PRESENCE_SCAN_TIMEOUT = "presence_scan_timeout"
    const val PRESENCE_SCAN_CANCELLED = "presence_scan_cancelled"
    const val PRESENCE_SCAN_SKIPPED = "presence_scan_skipped"
    const val PRESENCE_SCAN_FAILED = "presence_scan_failed"

    const val OWNER_GRANTED = "connection_owner_granted"
    const val OWNER_DENIED = "connection_owner_denied"
    const val OWNER_RELEASED = "connection_owner_released"

    const val CONNECT_INTENT_CREATED = "connect_intent_created"
    const val CONNECT_INTENT_CLEARED = "connect_intent_cleared"
    const val AUTO_CONNECT_PROMOTED = "auto_connect_promoted"
    const val AUTO_CONNECT_SKIPPED = "auto_connect_skipped"
    const val AUTO_START_ARMED = "auto_start_armed"
    const val AUTO_START_TRIGGERED = "auto_start_triggered"
    const val AUTO_START_SKIPPED = "auto_start_skipped"
    const val ALTERNATIVE_HINT_OFFERED = "alternative_hint_offered"
    const val ALTERNATIVE_HINT_ACCEPTED = "alternative_hint_accepted"
    const val ALTERNATIVE_HINT_DISMISSED = "alternative_hint_dismissed"

    const val PAUSE_STARTED = "connection_pause_started"
    const val PAUSE_CLEARED = "connection_pause_cleared"
    const val PAUSE_EXPIRED = "connection_pause_expired"
    const val PAUSE_BLOCKED = "connection_pause_blocked"

    const val SERVICE_STARTED = "connection_service_started"
    const val SERVICE_PROMOTED_FOREGROUND = "connection_service_promoted_foreground"
    const val SERVICE_DEMOTED_BACKGROUND = "connection_service_demoted_background"
    const val SERVICE_STOPPED = "connection_service_stopped"

    const val FOREGROUND_WORK_ACQUIRED = "foreground_work_acquired"
    const val FOREGROUND_WORK_RELEASED = "foreground_work_released"
    const val BACKGROUND_TASK_STARTED = "background_task_started"
    const val BACKGROUND_TASK_ENDED = "background_task_ended"
    const val BACKGROUND_TASK_EXPIRED = "background_task_expired"

    const val BOARD_SELECTED = "board_selected"
    const val BOARD_LINK_PERSISTED = "board_link_persisted"
    const val BOARD_LINK_FAILED = "board_link_failed"

    const val RIDE_SUMMARY_PREPARED = "ride_summary_prepared"
    const val RIDE_SUMMARY_NOTIFIED = "ride_summary_notified"
    const val RIDE_SUMMARY_SKIPPED = "ride_summary_skipped"
}

/** Who owns connection work, in the precedence order of ADR 0035. */
object ConnectionTraceOwner {
    const val BOARD_SESSION = "board_session"
    const val CONNECT_INTENT = "connect_intent"
    const val AUTO_START = "auto_start"
    const val AUTO_CONNECT = "auto_connect"
    const val ALTERNATIVE_HINT = "alternative_hint"
    const val ADD_BOARD_SCAN = "add_board_scan"
    const val BOARD_PROBE = "board_probe"
    const val NONE = "none"
}

/** Why the workflow exists. Set once at [ConnectionTrace.start]. */
object ConnectionTraceOrigin {
    const val FOREGROUND_ENTRY = "foreground_entry"
    const val EXPLICIT_CONNECT = "explicit_connect"
    const val AUTO_START_WAKE = "auto_start_wake"
    const val ALTERNATIVE_HINT_SWITCH = "alternative_hint_switch"
    const val BOARD_LINKED = "board_linked"
    const val ADD_BOARD_SCAN = "add_board_scan"
    const val BOARD_PROBE = "board_probe"
    const val RECONNECT = "reconnect"
    const val MANUAL_DISCONNECT = "manual_disconnect"
    const val END_RIDE = "end_ride"
    const val APP_EXIT = "app_exit"
    const val TASK_REMOVED = "task_removed"
    const val RIDE_FINALIZED = "ride_finalized"
}

/** Field names. Later slices reuse these instead of inventing new keys. */
object ConnectionTraceField {
    const val WORKFLOW_ID = "workflow_id"
    const val WORKFLOW_ORIGIN = "workflow_origin"
    const val WORKFLOW_OWNER = "workflow_owner"
    const val WORKFLOW_STARTED_AT = "workflow_started_at"
    const val ELAPSED_MS = "elapsed_ms"

    const val BOARD_ID = "board_id"
    const val BLE_ID = "ble_id"
    const val BOARD_NICKNAME = "board_nickname"

    const val DECISION = "decision"
    const val REASON = "reason"
    const val OWNER_PREVIOUS = "owner_previous"
    const val OWNER_REQUESTED = "owner_requested"

    const val DEADLINE_MS = "deadline_ms"
    const val DEADLINE_AT = "deadline_at"
    const val ATTEMPT = "attempt"

    const val SCAN_PURPOSE = "scan_purpose"
    const val OBSERVATION_COUNT = "observation_count"
    const val RSSI = "rssi"

    const val PAUSE_SOURCE = "pause_source"
    const val PAUSED_UNTIL = "paused_until"

    const val AUTO_CONNECT_ENABLED = "auto_connect_enabled"
    const val AUTO_START_ENABLED = "auto_start_enabled"
    const val BLUETOOTH_ENABLED = "bluetooth_enabled"
    const val PERMISSION_GRANTED = "permission_granted"
    const val APP_FOREGROUND = "app_foreground"
    const val SERVICE_STATE = "service_state"
    const val FOREGROUND_WORK = "foreground_work"

    const val RIDE_ID = "ride_id"
    const val PLATFORM_ERROR_CODE = "platform_error_code"
    const val PLATFORM_ERROR_DOMAIN = "platform_error_domain"
}

/** Value of [ConnectionTraceField.DECISION]. */
object ConnectionTraceDecision {
    const val GRANTED = "granted"
    const val DENIED = "denied"
    const val DEFERRED = "deferred"
    const val SKIPPED = "skipped"
    const val COMPLETED = "completed"
    const val TIMEOUT = "timeout"
    const val CANCELLED = "cancelled"
    const val FAILED = "failed"
}

/** Terminal reason names. Value of [ConnectionTraceField.REASON]. */
object ConnectionTraceReason {
    const val MATCHED = "matched"
    const val NO_LINKED_BOARDS = "no_linked_boards"
    const val NO_BOARD_LINK = "no_board_link"
    const val NO_SELECTED_BOARD = "no_selected_board"
    const val BOARD_NOT_PRESENT = "board_not_present"
    const val BLUETOOTH_DISABLED = "bluetooth_disabled"
    const val PERMISSION_MISSING = "permission_missing"
    const val SCANNER_UNAVAILABLE = "scanner_unavailable"
    const val SCANNER_BUSY = "scanner_busy"
    const val AUTO_CONNECT_DISABLED = "auto_connect_disabled"
    const val AUTO_START_DISABLED = "auto_start_disabled"
    const val CONNECTION_PAUSED = "connection_paused"
    const val HIGHER_PRIORITY_OWNER = "higher_priority_owner"
    const val SESSION_ALREADY_ACTIVE = "session_already_active"
    const val CONNECT_INTENT_ACTIVE = "connect_intent_active"
    const val USER_CANCELLED = "user_cancelled"
    const val STOP_SEARCH = "stop_search"
    const val DEADLINE_EXPIRED = "deadline_expired"
    const val MANUAL_DISCONNECT = "manual_disconnect"
    const val END_RIDE = "end_ride"
    const val APP_EXIT = "app_exit"
    const val TASK_REMOVED = "task_removed"
    const val AUTO_CLOSE = "auto_close"
    const val MECHANICAL_TEARDOWN = "mechanical_teardown"
    const val PROBE_CANCELLED = "probe_cancelled"
    const val PLATFORM_ERROR = "platform_error"
    const val RIDE_SUMMARY_DISABLED = "ride_summary_disabled"
    const val RIDE_NOT_ELIGIBLE = "ride_not_eligible"
    const val ALREADY_NOTIFIED = "already_notified"
}

/**
 * One correlated connection workflow. Pass the handle across layers — lifecycle → scanner →
 * connection → service → recording — so the correlation id survives handoff.
 */
class ConnectionWorkflow internal constructor(
    private val reporter: DiagnosticReporter,
    val workflowId: String,
    val origin: String,
    private val startedAtMs: Long,
    owner: String,
) {
    @Volatile
    var owner: String = owner
        private set

    /** Record a new owner for subsequent child events. Returns the same handle for chaining. */
    fun handoff(owner: String): ConnectionWorkflow {
        this.owner = owner
        return this
    }

    fun event(eventName: String, fields: Map<String, Any?> = emptyMap()) {
        reporter.capture(eventName, baseFields() + ConnectionTrace.sanitize(fields))
    }

    fun finish(decision: String, reason: String, fields: Map<String, Any?> = emptyMap()) {
        event(
            ConnectionTraceEvent.WORKFLOW_FINISHED,
            mapOf(
                ConnectionTraceField.DECISION to decision,
                ConnectionTraceField.REASON to reason,
            ) + fields,
        )
    }

    private fun baseFields(): Map<String, Any?> = mapOf(
        ConnectionTraceField.WORKFLOW_ID to workflowId,
        ConnectionTraceField.WORKFLOW_ORIGIN to origin,
        ConnectionTraceField.WORKFLOW_OWNER to owner,
        ConnectionTraceField.WORKFLOW_STARTED_AT to startedAtMs,
        ConnectionTraceField.ELAPSED_MS to ConnectionTrace.now() - startedAtMs,
    )
}

object ConnectionTrace {
    private val SENSITIVE_MARKERS = listOf(
        "auth",
        "credential",
        "jwt",
        "password",
        "payload",
        "pin",
        "secret",
        "session_token",
        "telemetry",
        "token",
    )

    /** Start a correlated workflow and emit `connection_workflow_started`. */
    fun start(
        context: Context,
        origin: String,
        owner: String = ConnectionTraceOwner.NONE,
        fields: Map<String, Any?> = emptyMap(),
    ): ConnectionWorkflow {
        val workflow = ConnectionWorkflow(
            reporter = DiagnosticReporter.get(context),
            workflowId = newOperationId(),
            origin = origin,
            startedAtMs = now(),
            owner = owner,
        )
        workflow.event(ConnectionTraceEvent.WORKFLOW_STARTED, fields)
        return workflow
    }

    /**
     * Rebuild a handle for a workflow that already started, so a layer that only received the
     * correlation id (service restart, notification action, JS intent) keeps emitting under it.
     */
    fun resume(
        context: Context,
        workflowId: String,
        origin: String,
        startedAtMs: Long,
        owner: String = ConnectionTraceOwner.NONE,
    ): ConnectionWorkflow = ConnectionWorkflow(
        reporter = DiagnosticReporter.get(context),
        workflowId = workflowId,
        origin = origin,
        startedAtMs = startedAtMs,
        owner = owner,
    )

    /** Contract-level exclusion of auth data, PINs, tokens, and telemetry payloads. */
    fun isSensitiveField(key: String): Boolean =
        SENSITIVE_MARKERS.any { key.contains(it, ignoreCase = true) }

    internal fun sanitize(fields: Map<String, Any?>): Map<String, Any?> =
        fields.filterKeys { !isSensitiveField(it) }

    internal fun now(): Long = System.currentTimeMillis()
}
